### PR TITLE
Fix Actions CI for MSVC

### LINKED ISSUE #616: CI fails with MSVC
Happened since #615; the MSVC CI job emits a bunch of warnings, originating from the C library itself (!?), whic @JL2210 could *not* reproduce on his machine. I should look into reproducing this, perhaps GitHub provides VM images for their CI machines?

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #616

### CODE REVIEW

**Comment by Rangi42** [comment]
One remaining error:

```
LINK : fatal error LNK1181: cannot open input file 'png16.lib' [D:\a\rgbds\rgbds\build\src\rgbgfx.vcxproj]
```

And one warning:

```
D:\a\rgbds\rgbds\src\link\object.c(640,23): warning C4090: 'function': different 'const' qualifiers [D:\a\rgbds\rgbds\build\src\rgblink.vcxproj]
```

**Comment by Rangi42** [comment]
@JL2210 Any idea what the matter is with `png16.lib`? (The log shows that `libpng16.lib` is created.)

**Comment by Rangi42** [comment]
@JL2210 @ISSOtm This [bug](https://github.com/microsoft/vcpkg/pull/9198) and [fix](https://github.com/microsoft/vcpkg/pull/9198/commits/09ee6e9ce82ec0d4a050b63a5006ff481b977c8a) look related (it's one of the only web search results for "png16.lib"). Patching libpng's own CMakeLists.txt so it creates the right path for MSVC.

**Comment by Rangi42** [comment]
Note that the MSCV builds broke between 0bb5efebfdc50024b3752ce2b82f4f1366a29c99 and cc4d455b8a827e03ea73554a6058a7888bf9e736, apparently not due to any change in rgbds. (That doesn't have the `LNK1181` error over `png16.lib`, but I think that's because it failed for other reasons (the ones this PR fixes so far) before even getting to that point.)

**Comment by Rangi42** [comment]
Despite the `png16.lib` issue, this clears up all the other errors and warnings cluttering the build log, so IMO it could be merged as-is. @ISSOtm

**Comment by ISSOtm** [comment]
> @JL2210 @ISSOtm This [bug](https://github.com/microsoft/vcpkg/pull/9198) and [fix](https://github.com/microsoft/vcpkg/pull/9198/commits/09ee6e9ce82ec0d4a050b63a5006ff481b977c8a) look related (it's one of the only web search results for "png16.lib"). Patching libpng's own CMakeLists.txt so it creates the right path for MSVC. I don't see how to do that for rgbds's CI build though.

I have two ideas to fix that:
- Avoid using `pkg_config` on Windows
- Manually perform a replace (`png16.lib` → `libpng16.lib`) if that doesn't work

I don't know nearly enough about Windows / CMake to have any idea if those would work...

File: src/asm/lexer.c [hunk 1]
```diff
@@ -47,9 +47,12 @@
 
 /* Neither MSVC nor MinGW provide `mmap` */
 #if defined(_MSC_VER) || defined(__MINGW32__)
-# include <windows.h>
-# include <fileapi.h>
-# include <winbase.h>
+# define WIN32_LEAN_AND_MEAN
```
Comment (ISSOtm) [inline src/asm/lexer.c:50]: What does this do?

File: include/asm/symbol.h [hunk 2]
```diff
@@ -117,7 +117,11 @@ void sym_SetExportAll(bool set);
 struct Symbol *sym_AddLocalLabel(char const *symName);
 struct Symbol *sym_AddLabel(char const *symName);
 struct Symbol *sym_AddAnonLabel(void);
-void sym_WriteAnonLabelName(char name[static MAXSYMLEN + 1], uint32_t ofs, bool neg);
+#ifdef _MSC_VER
+void sym_WriteAnonLabelName(char *buf, uint32_t ofs, bool neg);
+#else
+void sym_WriteAnonLabelName(char buf[static MAXSYMLEN + 1], uint32_t ofs, bool neg);
```
Comment (ISSOtm) [inline include/asm/symbol.h:123]: This warrants further discussion, which I'll take to the main issue thread.

File: CMakeLists.txt [hunk 3]
```diff
@@ -31,7 +31,7 @@ option(TRACE_PARSER "Trace parser execution" OFF)
 option(TRACE_LEXER "Trace lexer execution" OFF)
 
 if(MSVC)
-  add_compile_options(/W1 /MP)
+  add_compile_options(/W1 /MP /wd5105)
```
Comment (ISSOtm) [inline CMakeLists.txt:34]: Is that flag the "ignore macro expansions producing `defined`" flag? If yes, a comment pointing out that it should be removed ASAP would be good.

**Review by ISSOtm — COMMENTED** [review]
Thank you for being the one putting the effort... ^^'

**Comment by ISSOtm** [comment]
So, apparently MSVC doesn't support `[static N]`. Thing is, `Type ptr[static 1]` is useful to point out that a pointer is non-nullable. (Thanks @aaaaaa123456789 for pointing that out!)

This would be useful to add to other functions, especially to indicate whether they should be checked for `NULL` failure, or they don't need to. Should we define a macro (`non_null(struct Symbol, ptr)` and/or `min_nb_elms(char, ptr, MAXSYMLEN + 1)`), or drop that idea entirely?

File: src/asm/lexer.c (hunk 1 above)
Comment (Rangi42) [reply src/asm/lexer.c:50]: `WIN32_LEAN_AND_MEAN` makes `windows.h` include less. Something it was including without that was causing a build error. (I don't remember exactly what, I was doing a lot of force-pushes attempting to not include windows.h at all before finding that worked.)

File: CMakeLists.txt (hunk 3 above)
Comment (Rangi42) [reply CMakeLists.txt:34]: Yes, I'll add a comment for that.

**Comment by Rangi42** [comment]
I had taken the `ifdef`/`else`/`endif` approach because it only applied to one function. If it's at all likely to be used more than once, I do think a macro would be more appropriate. It would also be the right place to document the MSVC lack of support.

How about:

```
#ifdef _MSC_VER
// Microsoft Visual Studio does not support [static N] from C99
#define MIN_NB_ELMS(arg, len) arg
#else
#define MIN_NB_ELMS(arg, n) arg[static n]
#endif

char MIN_NB_ELMS(buf, MAXSYMLEN + 1)
```

**Comment by ISSOtm** [comment]
```diff
-#define MIN_NB_ELMS(arg, len) arg
+#define MIN_NB_ELMS(arg, len) *arg
```
, but other than that I agree. We'd put it in `platform.h` with the rest of the compatibility hacks.

**Comment by aaaaaa123456789** [comment]
> ```diff
> -#define MIN_NB_ELMS(arg, len) arg
> +#define MIN_NB_ELMS(arg, len) *arg
> ```
> 
> , but other than that I agree. We'd put it in `platform.h` with the rest of the compatibility hacks.

`#define MIN_NB_ELEMS(arg, len) arg[]` to avoid precedence issues.

**Comment by ISSOtm** [comment]
After doing some tests on my end, I'm OK with `/wd5105` and a comment. I don't like it, but that's better than keeping the logs to their current size.

Additionally, I'd like to note that [additional "lean" defines](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2065?view=msvc-160#example-missing-header-file) allow the build to [succeed even with the warning still active](https://github.com/ISSOtm/rgbds/runs/1641147385?check_suite_focus=true#step:6:67), fwiw.

**Comment by ISSOtm** [comment]
[Microsoft's docs](https://docs.microsoft.com/en-us/cpp/build/walkthrough-compile-a-c-program-on-the-command-line?view=msvc-160#next-steps) note that C11/C17 has to be enabled specially via `/std:c11`/`std:c17`. [We set appropriate defines](https://github.com/gbdev/rgbds/blob/77279984a53054436450ea251343515bfdbf25b1/CMakeLists.txt#L70-L71), but I don't know if CMake got support for MSVC's options (including whether the version on GitHub Actions does). I wasn't able to get CI to echo back the commands it's running, either.

**Comment by Rangi42** [comment]
Added comments, the `MIN_NB_ELMS` macro, `/std:c11` (matches `-std=gnu11`), and the png16.lib fix.

File: include/platform.h [hunk 4]
```diff
@@ -39,4 +39,11 @@
 # define SSIZE_MAX INT_MAX
 #endif
 
+/* MSVC doesn't support `[static N]` for array arguments from C99 */
+#ifdef _MSC_VER
+# define MIN_NB_ELMS(N)
+#else
+# define MIN_NB_ELMS(N) static N
```
Comment (ISSOtm) [inline include/platform.h:46]: Clever impl! Might be better to add parens around that `N`, though?

**Review by ISSOtm — APPROVED** [review]

**Comment by ISSOtm** [comment]
Time to rebase everything!

**Pull request merged**